README.md

```diff
@@ -2,27 +2,179 @@
+* docker installed. (docker-ce-18.03.1.ce-1.fc27.x86_64.rpm package)
+* the `raid1` and `dm_raid` kernel modules must be loaded. If it isn't, run `modprobe raid1 dm_raid` as root.
+
+Then run
+
+```bash
+make sudo-test
+```
+
+While developing in only one package it is simpler and faster to run only certain tests.
+
+For example, if you're adding new functionality to the `./pkg/lvm` package you can do
+:
+```bash
+cd ./pkg/lvm
+go test -c -i . && sudo ./lvm.test -test.v -test.run=TestMyNewFeature
+```
+
+
+## How does this plugin map to the CSI specification?
+
+This plugin is a CSI-compliant wrapper around the normal `lvm2` command-line utilities.
+These include `pvcreate`, `vgcreate`, `lvcreate`, etc.
+
+This plugin implements the "headless, unified plugin" design given in the CSI specification.
+
+In particular, the CSI specification lists the architecture as follows:
+
+```
+                            CO "Node" Host(s)
++-------------------------------------------+
+|                                           |
+|  +------------+           +------------+  |
+|  |     CO     |   gRPC    | Controller |  |
+|  |            +----------->    Node    |  |
+|  +------------+           |   Plugin   |  |
+|                           +------------+  |
+|                                           |
++-------------------------------------------+
+
+Figure 3: Headless Plugin deployment, only the CO Node hosts run
+Plugins. A unified Plugin component supplies both the Controller
+Service and Node Service.
+```
+
+Every instance of this plugin controls a single LVM2 volume group (VG).
+Every CSI volume corresponds to a LVM2 logical volume (LV).
+The CSI RPCs map to command-line utility invocations on LVM2 physical volumes (PV), a LVM2 volume group (VG) or LVM2 logical volumes (LV).
+For the the exact command-line invocations read the source code, starting at https://github.com/mesosphere/csilvm/blob/master/pkg/csilvm/server.go.
+
+
+## Project structure
+
+This project is split into `./pkg` and `./cmd` directories.
+
+The `./pkg` directory contains logic that may be used from unit tests.
+
+The `./cmd` directory contains commands that interface between the environment (e.g., parsing command-line options, reading environment variables, etc.) and the logic contained in the `./pkg` directory.
+
+The `./pkg` directory is split into the `./pkg/lvm` and `./pkg/csilvm` packages.
+
+The `./pkg/lvm` package provides a Go wrapper around LVM2 command-line utilities and actions.
+
+The `./pkg/csilvm` package includes all the CSI-related logic and translates CSI RPCs to `./pkg/lvm` function calls.
+
+The `./pkg/csilvm/server.go` file should serve as your entrypoint when reading the code as it includes all the CSI RPC endpoints.
 
-Logical volumes are named according to the following pattern
-`<volume-group-name>_<logical-volume-name>`.
 
-## Runtime dependencies
+## Deployment
+
+The plugin builds to a single executable binary called `csilvm`.
```

Comment:
Does the binary has any dependency, or is it self contained (i.e., static)?

Reply:
It is _not_ static in that it doesn't use a pure-go networking stack and instead links dynamically against libc for networking. Specifically,

```
$ ldd ./csilvm
	linux-vdso.so.1 (0x00007fffc96a7000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f98146cb000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f9814315000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f98148e9000)
```

If we use a pure-go networking stack one could generate a static binary:
```
$ go build -a -tags netgo ./cmd/csilvm
$ ldd ./csilvm
	not a dynamic executable
```

We can add a Make target for building a static binary.